mpy-cross: allow using stdin and stdout instead of files

arbrauns opened the pull request:
Specifying `-` for the input or output file name uses stdin and stdout instead.

arbrauns committed:
mpy-cross: Allow specifying source files starting with -.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns committed:
mpy-cross: Allow reading from stdin and writing to stdout.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

dlech commented on mpy-cross/main.c:
```diff
@@ -77,17 +90,22 @@ STATIC int compile_and_save(const char *file, const char *output_file, const cha
+            vstr_t vstr;
+            vstr_init(&vstr, 16);
+            if (output_file == NULL) {
+                vstr_add_str(&vstr, file);
+                vstr_cut_tail_bytes(&vstr, 2);
+                vstr_add_str(&vstr, "mpy");
```
Won't this make the output file `-.mpy` if the input file is `-`?

It might be better to make the default to output to stdout if the input is stdin and no output argument is given.

arbrauns replied:
Indeed, I've changed it accordingly.

dlech commented on mpy-cross/main.c:
```diff
@@ -309,6 +328,8 @@ MP_NOINLINE int main_(int argc, char **argv) {
                 } else {
                     return usage(argv);
                 }
+            } else if (strcmp(argv[a], "--") == 0) {
```
Am I reading this right and `--` is required in order to use `-` as the input file?

IMHO, it would be nicer to leave out `--` since we don't really need it for anything else and add a case instead for matching `-` by itself.

arbrauns replied:
`--` is a common convention, and is required for any filenames starting with `-`. Not a particularly common use case, I admit, but essential for script hardening.

Another common convention is defaulting to stdin if no input file is given, maybe that would be a good alternative to having to specify `-- -`?

dlech replied:
> Another common convention is defaulting to stdin if no input file is given

That sounds fine to me.

dpgeorge replied:
I think it would be good to also add a case to handle `-` explicitly, even when `option_parsing_active` is true.  (Eg `diff` allows passing `-` directly.)

arbrauns replied:
Done.

dlech reviewed (commented):
I think this is a nice addition for using `mpy-cross` programatically to avoid having to manage temporary files.

arbrauns committed:
mpy-cross: When reading from stdin, write output to stdout.

Unless -o is given, output defaults to stdout unless a source file is
given (in which case the source file name is used to derive an output
file name).

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

dlech commented:
I think this should be split into two commits (and perhaps even better, to pull requests). One for adding stdin support and one for adding `--` as those are independent features.

dpgeorge commented on mpy-cross/qstrdefsport.h:
```diff
@@ -1 +1,3 @@
 // qstrs specific to this port
+// *FORMAT-OFF*
+Q(<stdin>)
```
This shouldn't be needed because it's in `py/qstrdefs.h`.

arbrauns replied:
It didn't work for me when I tried it (it expanded to `_lt_stdin_gt_`), having it specified twice shouldn't hurt, right?

dpgeorge replied:
I just tried it without this line added, and it works for me.  Builds without error and then running it looks correct:
```
$ echo -n | ./build/mpy-cross - | xxd 
00000000: 4d06 001f 0200 0e3c 7374 6469 6e3e 000f  M......<stdin>..
00000010: 2800 0201 5163                           (...Qc
```

arbrauns replied:
Alright, not sure what that was then, I've removed it.

dpgeorge commented on mpy-cross/main.c:
```diff
@@ -322,8 +344,8 @@ MP_NOINLINE int main_(int argc, char **argv) {
-        mp_printf(&mp_stderr_print, "no input file\n");
-        exit(1);
+        // default to stdin
+        input_file = "-";
```
I prefer the existing behaviour of giving an error if no files are specified (printing "no input file").  Because that matches gcc/clang, and they are the closest unix tool to mpy-cross.

arbrauns replied:
Done.

arbrauns commented:
> I think this should be split into two commits (and perhaps even better, to pull requests). One for adding stdin support and one for adding `--` as those are independent features.

They already are separate commits. The first commit adds `--` support, the second allows specifying `-` for stdin/stdout, the third makes output default to stdout if reading from stdin, and the fourth makes stdin the default if no source file is specified.

arbrauns committed:
mpy-cross: Allow specifying stdin as input without --.

This way, a bare `-` is never interpreted as an option, even before
`--`. Filenames starting with `-` still need to be put after `--`.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

dpgeorge commented:
Thanks for updating, this looks good now.

dpgeorge commented:
And thanks for the contribution (with nice clean commits)!
